helper/resource/testing_new.go

```diff
@@ -0,0 +1,615 @@
+	tfjson "github.com/hashicorp/terraform-json"
+	"github.com/hashicorp/terraform-plugin-sdk/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/internal/addrs"
+	"github.com/hashicorp/terraform-plugin-sdk/internal/tfdiags"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+)
+
+func shimAttributeValues(flatmap map[string]string, currentKey string, value interface{}) {
+	switch v := value.(type) {
+	case nil:
+		// omit the entry altogether
+	case bool:
+		flatmap[currentKey] = strconv.FormatBool(v)
+	case float64:
+		flatmap[currentKey] = strconv.FormatFloat(v, 'f', -1, 64)
+	case string:
+		flatmap[currentKey] = v
+	case map[string]interface{}:
+		if currentKey != "" {
+			currentKey += "."
+		}
+		for key, val := range v {
+			shimAttributeValues(flatmap, fmt.Sprintf("%s%s", currentKey, key), val)
+		}
+		flatmap[currentKey+"%"] = strconv.Itoa(len(v))
+	case []interface{}:
+		if currentKey != "" {
+			currentKey += "."
+		}
+		for i, val := range v {
+			shimAttributeValues(flatmap, fmt.Sprintf("%s%d", currentKey, i), val)
+		}
+		flatmap[currentKey+"#"] = strconv.Itoa(len(v))
+	default:
+		panic("Unknown json type")
+	}
+}
+
+func shimStateModule(state *terraform.State, newModule *tfjson.StateModule) error {
+	var path addrs.ModuleInstance
+	var diags tfdiags.Diagnostics
+	if newModule.Address == "" {
+		path = addrs.RootModuleInstance
+	} else {
+		path, diags = addrs.ParseModuleInstanceStr(newModule.Address)
+		if diags.HasErrors() {
+			return diags.Err()
+		}
+	}
+
+	mod := state.AddModule(path)
+	for _, res := range newModule.Resources {
+		resState := &terraform.ResourceState{
+			Provider: res.ProviderName,
+			Type:     res.Type,
+		}
+
+		flatmap := make(map[string]string)
+		shimAttributeValues(flatmap, "", res.AttributeValues)
+
+		if _, exists := flatmap["id"]; !exists {
+			return errors.New("attributes had no id")
+		}
+
+		resState.Primary = &terraform.InstanceState{
+			Tainted:    res.Tainted,
+			ID:         flatmap["id"],
+			Attributes: flatmap,
+			Meta: map[string]interface{}{
+				"schema_version": res.SchemaVersion,
+			},
+		}
+
+		resState.Dependencies = res.DependsOn
+
+		idx := ""
+		switch v := res.Index.(type) {
+		case int:
+			idx = fmt.Sprintf(".%d", v)
+		case string:
+			idx = "." + v
+		}
+
+		mod.Resources[res.Address+idx] = resState
+	}
+
+	for _, child := range newModule.ChildModules {
+		if err := shimStateModule(state, child); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func shimTFJson(jsonState *tfjson.State) (*terraform.State, error) {
+	state := terraform.NewState()
+	state.TFVersion = jsonState.TerraformVersion
+	if jsonState.Values == nil {
+		// the state is empty
+		return state, nil
+	}
+
+	if err := shimStateModule(state, jsonState.Values.RootModule); err != nil {
+		return nil, err
+	}
+
+	// shimming of lists and maps might be incorrect
+	for key, output := range jsonState.Values.Outputs {
```

Comment:
I really had no clue if this part "worked", would be best to test against a provider that makes checks against outputes, this is common in datasource tests, one that checks against a list or map would be a great candidate (I wasn't concerned much about the string case)